### PR TITLE
Webhooks

### PR DESCRIPTION
**Pull request by Zomatree**

### CODE REVIEW

**Comment by itzTheMeow** [comment]
could the author ID be the webhook ID? and the `webhook` property holds the additional webhook info

**Review by arktn — COMMENTED** [review]
Why would message author be optional? I would think something along the lines of @itzTheMeow's comment would be better to work with. Leave the author id out and just use the webhook id.

**Review by arktn — APPROVED** [review]

**Review by arktn — APPROVED** [review]
.

**Comment by Zomatree** [comment]
Both ways is still a breaking change, I would prefer to properly separate them out, having to check other keys to find out how to parse a different item is not a good way to handle it.

**Comment by itzTheMeow** [comment]
you still have to check that the webhook key exists (or that the author id doesn't). also why not send the webhook name/avatar with the message instead of making the user fetch it separately?

File: rfcs/rfc-0001-webhooks.md [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+
+```rust
+struct Webhoook {
+  // Unique Id
+  id: String,
+
+  // The name of the webhook - 1 to 32 chars
+  name: String,
+
+  // The avatar of the webhook, it it has one
```
Comment (itzTheMeow) [inline rfcs/rfc-0001-webhooks.md:46]:
```suggestion
  // The avatar of the webhook, if it has one
```
minor typo, not sure if it matters

**Comment by itzTheMeow** [comment]
is there going to be a way to reset the webhook token without recreating it? or does that seem unnecessary

**Comment by insertish** [comment]
As per internal group chat, if the webhook is actually attached to a message, it should use the following struct:
```rust
struct MessageWebhook {
  // Id of Webhook
  id: String,
  // The name of the webhook - 1 to 32 chars
  name: String,
  // The id of the avatar of the webhook, if it has one
  avatar: Option<String>,
}
```

**Comment by Andre601** [comment]
One thing I was wondering about. Given the github integration, why not also support other remote git repo hosts like GitLab, BitBucket, Codeberg, etc, if they allow webhook sending?

Maybe even have two general paths `/webhooks/<target>/<token>/gitea` and `/webhooks/<target>/<token>/forgejo` to allow a more general support for self-hosted repos using Gitea or Forgejo respectively (I assume Forgejo utilizes the same webhook mechanics as Gitea but could be good to keep it separate, just in case), allowing a lot more repo hosts to be supported (Would eliminate the need to support Codeberg as it runs on Forgejo).

**Comment by Andre601** [comment]
Also, this is unrelated to this RFC in particular, but to RFC in general: Maybe re-structure the shown API routes a bit by having a code highlighting (`http`) applied and also align the paths for easier readability?

See my previous comment for a working example.
This imo improves readability a bit by distinguishing the method and path from each other.

**Comment by Zomatree** [comment]
> One thing I was wondering about. Given the GitHub integration, why not also support other remote git repo hosts like GitLab, BitBucket, Codeberg, etc, if they allow webhook sending?

I mentioned in the future ideas section on expanding the supported platforms, in the future I think it would be good to add more in another RFC, for the core webhooks RFC just GitHub support will be enough.

I cant see why a edit message route would be needed for webhooks, as all you should need is just creating messages, if you can think of some good examples on where you would need it then I might consider it.

**Comment by Rexogamer** [comment]
It might be useful if someone, say, edits an issue/comment?

**Comment by Rexogamer** [comment]
> is there going to be a way to reset the webhook token without recreating it? or does that seem unnecessary

For what it's worth this also seems like a good idea to me

**Comment by Zomatree** [comment]
> It might be useful if someone, say, edits an issue/comment? For what it's worth I think Discord lets you edit webhook messages, although if so they don't use this for the GitHub integration so hmm

webhooks are stateless in how they are implemented so we have no way of mapping the edit to the correct message, i dont think its worth maintaining a state for this.

**Comment by Andre601** [comment]
> I cant see why a edit message route would be needed for webhooks, as all you should need is just creating messages, if you can think of some good examples on where you would need it then I might consider it.

The only case I could give is something similar to Discord's announcement feature, allowing to update a previously crossposted message.
Tho, this may be a thing for the bridge feature instead.

**Comment by insertish** [comment]
> Tho, this may be a thing for the bridge feature instead.

Bridges as they currently exist don't use webhooks and instead use masquerades, and I don't think there's any benefit to switching to webhooks. "Native bridges" as they would be implemented later, would not have anything to do with webhooks either.

**Comment by insertish** [comment]
I think this is pretty much ready to merge now?
Only thing that I think might be unaccounted for is what kind of channels webhooks can be created on, I would probably deny Saved Messages, Voice Channels, and DMs but not group DMs (since those have settings).

**Comment by xXBuilderBXx** [comment]
Webhooks will have some interesting uses for Revolt but definitely could be better than other platforms.

Maybe a good idea would be to add a seperate permission set for webhooks so you can restrict masquerade so the webhook can't change name/avatar and only use default one, send embeds, use external emojis from other servers (not sure if it can check) and other stuff in the future.

**Comment by xXBuilderBXx** [comment]
Most services also have a /slack webhook type that is popular to use by many services, would there be support for this.

**Comment by Andre601** [comment]
> Webhooks will have some interesting uses for Revolt but definitely could be better than other platforms.
> 
> Maybe a good idea would be to add a seperate permission set for webhooks so you can restrict masquerade so the webhook can't change name/avatar and only use default one, send embeds, use external emojis from other servers (not sure if it can check) and other stuff in the future.

Agree on this.
I hate Discord's system where Webhooks inherit permissions from the default (everyone) role instead of allowing a different set of permissions to be applied.
Can make it risky where you want to allow a webhook to mention roles via the "Tag everyone, here and all roles" permissions, but with the channel being accessible to everyone with write perms...

**Comment by insertish** [comment]
Webhooks will be available on staging soon for testing, some caveats which need to be addressed:

- Webhooks ignore all permissions currently.
- There is no way to restrict mentions on a message.

**Comment by Zomatree** [comment]
This RFC is now in its final comment period, if no other discussions is needed, this RFC can be merged.

**Comment by Andre601** [comment]
There is one thing I was wondering about.
I didn't research it enough, so sorry if there is a clear answer to this...

Does Revolt have a base-level check for malicious URLs using a database when messages are send to prevent such sending.
I ask because if that was the case, would it also extend to webhook messages?

**Comment by insertish** [comment]
> Does Revolt have a base-level check for malicious URLs using a database when messages are send to prevent such sending. I ask because if that was the case, would it also extend to webhook messages?

Currently no, however if it were to be added it would automatically be available for webhooks since they share the message sending code.

**Pull request merged**